cairo/src/utils/dict.cairo

```diff
@@ -136,3 +137,83 @@ func hashdict_write{poseidon_ptr: PoseidonBuiltin*, dict_ptr: DictAccess*}(
     let dict_ptr = dict_ptr + DictAccess.SIZE;
     return ();
 }
+
+func hashdict_finalize{range_check_ptr}(
+    dict_ptr_start: DictAccess*,
+    dict_ptr: DictAccess*,
+    original_mapping_start: DictAccess*,
+    original_mapping_end: DictAccess*,
+    should_merge: felt,
+) -> (DictAccess*, DictAccess*) {
+    alloc_locals;
+
+    if (should_merge == FALSE) {
+        let (squashed_dict_start: DictAccess*) = alloc();
+        let (squashed_dict_end) = squash_dict(dict_ptr_start, dict_ptr, squashed_dict_start);
+        let (prev_values_start, prev_values_end) = prev_values(
+            squashed_dict_start, squashed_dict_end
+        );
+        tempvar dict_ptr_start = prev_values_start;
+        tempvar dict_ptr = prev_values_end;
+        tempvar range_check_ptr = range_check_ptr;
+    } else {
+        tempvar dict_ptr_start = dict_ptr_start;
+        tempvar dict_ptr = dict_ptr;
+        tempvar range_check_ptr = range_check_ptr;
+    }
+    let dict_ptr_start = cast([ap - 3], DictAccess*);
+    let dict_ptr = cast([ap - 2], DictAccess*);
+    let range_check_ptr = [ap - 1];
+
+    if (cast(original_mapping_end, felt) == 0) {
+        // No parent mapping, just return the current dict.
+        return (dict_ptr_start, dict_ptr);
+    }
+
+    let segment_to_merge = dict_ptr_start;
+    let len_to_merge = dict_ptr - dict_ptr_start;
+    let dst_ptr = original_mapping_end;
+    memcpy(dst_ptr, segment_to_merge, len_to_merge);
```

Comment:
tbh I find it more explicit than doing all these intermediate namings

```suggestion
    memcpy(original_mapping_end, dict_ptr_start, dict_ptr - dict_ptr_start);
```

Reply:
you can `tempvar len = dict_ptr - dict_ptr_start` to reuse it in both the `memcpy` and the `return`